src/components/tooltip/Marker.js

```diff
@@ -735,119 +108,257 @@ class MarkerTooltipContents extends React.PureComponent<Props> {
+      return null;
+    }
+
+    // We only know the duration if it's complete.
+    const duration = marker.incomplete
+      ? 'unknown duration'
+      : formatTimestamp(marker.end - marker.start, 3, 1);
+
+    return <div className="tooltipTiming">{duration}</div>;
+  }
+
+  _maybeRenderBacktrace() {
+    const { marker, thread, implementationFilter } = this.props;
+    const { data, start } = marker;
+    if (data && 'cause' in data && data.cause) {
+      const { cause } = data;
+      const causeAge = start - cause.time;
+      return (
+        <div className="tooltipDetailsBackTrace" key="backtrace">
+          {data.type === 'Styles' || marker.name === 'Reflow' ? (
+            <h2 className="tooltipBackTraceTitle">
+              First invalidated {formatNumber(causeAge)}ms before the flush, at:
+            </h2>
+          ) : null}
+          <Backtrace
+            maxHeight="30em"
+            stackIndex={cause.stack}
+            thread={thread}
+            implementationFilter={implementationFilter}
+          />
+        </div>
+      );
+    }
+    return null;
+  }
+
+  _maybeRenderNetworkPhases() {
     const {
-      marker,
-      className,
-      threadName,
-      thread,
-      implementationFilter,
+      marker: { data },
       zeroAt,
-      threadIdToNameMap,
     } = this.props;
+    if (data && data.type === 'Network') {
+      return <TooltipNetworkMarkerPhases payload={data} zeroAt={zeroAt} />;
+    }
+    return null;
+  }
 
-    const url = this._getUrl(marker);
-    const details = getMarkerDetails(
-      marker,
-      thread,
-      implementationFilter,
-      zeroAt
-    );
-    const threadDetail = this._getThreadDetail(
-      marker,
-      threadName,
-      threadIdToNameMap
-    );
+  /**
+   * Often-times component logic is split out into several different components. This
+   * is really helpful for interactive components, as each list of Props serves as
+   * a piece of memoization that can help with not over-rendering.
+   *
+   * This component is a little bit different, as it will only be rendered once per
+   * marker, as it's hovered. In addition, most of the complexity is in generating
+   * a list of TooltipDetailComponent, which are label/value pairs. Many parts
+   * of the tooltip are controlled by the Marker Schema, however there are some
+   * custom implementations for various markers that include some complexity.
+   * The general practice here is to provide functions of the signature:
+   *
+   * Payload => TooltipDetailComponent[]
+   *
+   * These TooltipDetailComponent are then combined with those generated by the
+   * Marker Schema.
+   *
+   * Finally, many sections are optional based on the data that is present. Note
+   * that these render function begin with "maybe", rather than embedded complicated
+   * ternaries everywhere. This leads to a style of render function that includes
+   * a short list of rendering strategies, in the order they appear.
+   */
```

Comment:
Nice comment!

Reply:
Just trying to justify my design decisions :D